### PR TITLE
Retrieve pin unpin and collect in Env.

### PR DESCRIPTION
**Pull request by AureliaDolo**

### CODE REVIEW

File: src/string_ptr.rs [hunk 1]
```diff
@@ -39,21 +44,38 @@ impl Write<String> for StringPtr {
...
+        // pin
+        let pin = export_asr!(fn_pin, env);
+        pin.call(&[Value::I32(offset.try_into().unwrap())])
+            .expect("Failed to call __pin");
+
         Ok(Box::new(StringPtr::new(offset)))
     }
 
-    fn write(&self, value: &str, env: &Env) -> anyhow::Result<()> {
+    fn write(&mut self, value: &str, env: &Env) -> anyhow::Result<Box<StringPtr>> {
```
Comment (AureliaDolo) [inline src/string_ptr.rs:56]: Not quite sure about making self mutable , but I did not see any way around

File: src/string_ptr.rs [hunk 2]
```diff
@@ -1,9 +1,17 @@
 use super::{Env, Memory, Read, Write};
-use std::convert::TryFrom;
+use std::convert::{TryFrom, TryInto};
 use wasmer::{Array, Value, WasmPtr};
 
 pub type StringPtr = WasmPtr<u16, Array>;
 
+macro_rules! export_asr {
```
Comment (yvan-sraka) [inline src/string_ptr.rs:7]: Nice macro ✨

File: src/string_ptr.rs (hunk 1 above)
Comment (yvan-sraka) [reply src/string_ptr.rs:56]: It completely made sense that a `write` needs a mutable reference on self!

**Review by yvan-sraka — APPROVED** [review]

File: src/string_ptr.rs (hunk 1 above)
Comment (AureliaDolo) [reply src/string_ptr.rs:56]: Indeed x)

File: src/string_ptr.rs (hunk 2 above)
Comment (AureliaDolo) [reply src/string_ptr.rs:7]: My first one ever :tophat:

**Comment by AureliaDolo** [comment]
@adrien-zinger as it's your repo, I will let you merge that :)

**Comment by adrien-zinger** [comment]
> @adrien-zinger as it's your repo, I will let you merge that :)

@AureliaDolo you can merge if Yvan approved!

**Pull request merged**